# SPQR - SPlit and conQueR for RNA structure prediction

SPQR is a coarse-grained representation and energy function for the prediction of the three-dimensional structure of RNA from the knowledge of its sequence [1].

A nucleotide is represented as two particles: an anisotropic one which stands for the nucleoside and a point particle for the phosphate group.
<p align="center">
<img src="https://github.com/srnas/spqr/blob/master/doc/src/rnacg.png" width="200">
<p>

## Features

### Interactions
Our energy function makes a geometrically detailed description of canonical and non-canonical base pairs, stacking interactions and even base-phosphate interactions! They are constructed as knowledge-based potentials inspired by the ESCORE function [2].

### Base degrees of freedom
Also, SPQR allows the definition of different glycosidic bond angle conformations: ANTI, HIGH-ANTI and SYN, and it is able to distinguish between sugar pucker conformations: C3' endo and C2' endo.

### Backmapping procedure
Due to the anisotropic representation of the nucleobases, SPQR allows the reintroduction of atomistic details in the predicted structures through steered-Molecular Dynamics [3].

SPQR has successfully been tested on several structures, including duplexes, hairpins and junctions [1].

### The code
SPQR is implemented in a versatile and clear, bug free C code which is available to download for free. The simulations can run constant temperature Monte Carlo simulations, and use Simulated Annealing and Simulated Tempering algorithms for finding the minimum energy conformations. It also includes a set of analysis tools for making the whole process easier.

We include here a small tutorial for the folding of a small tetraloop hairpin, and a user manual.

# Getting started

## Installing

After downloading the source, go to the `spqr/src` directory and run
```
autoconf
./configure
make
make install
```

The binary files will be stored in the `spqr/bin` directory and the tools, in the `spqr/tools` file. A more detailed manual is stored in `spqr/doc`.

## Quick energy minimization

A simple minimization can be run on a pdb file containing a single-strand of RNA. Copy the `SPQR_REFINE` script into any directory and run
```
./SPQR_REFINE -i <PDBFILE> 
```
The result will create the folder refSPQR<PDBFILE>. Optionally, the name can be specified with the `-o <OUTPUTNAME>` option.
The folder will contain a pdb file in SPQR format, an all-atom reconstruction and a .mc file, which can be used for further SPQR simulations with full precision (recommended). The all-atom reconstruction is suitable for MD simulations as well.
The minimization of energy is a good starting point for evaluating the SPQR energy, which is contained in the `.ene` file in the working directory. In addition, it removes clashes and fixes broken bonds, which might be of help when dealing with structures generated by fragment assembly methods, such as the ones generated by Ernwin [4].
If the secondary structure elements such as hairpins, stems and internal loops are entangled, these artifacts can also be removed automatically [5]. In this case, the secondary structure must be provided in a fasta file containing three lines : a comment line, the sequence and the secondary structure in Vienna format, without pseudoknots. The option `-t <SECONDARYSTRUCTURE>` takes care of this.
Other options can be found in the user's guide and the help message displayed with the `-h` flag.










## References
[1] S. Poblete, S. Bottaro and G. Bussi, A nucleobase-centric coarse-grained model for structure prediction of RNA motifs. _Nucleic Acids Res._ 46(4), 1674-1683 (2018).  [Link] https://doi.org/10.1093/nar/gkx1269 

[2] S. Bottaro, F. Di Palma and G. Bussi, The role of nucleobase interactions in RNA structure and dynamics. _Nucleic Acids Res._ 42(21), 13306-13314 (2014).

[3] S. Poblete, S. Bottaro and G. Bussi, Effects and limitations of a nucleobase-driven backmapping procedure for nucleic acids using steered molecular dynamics. _Biophys. Biochem. Res. Comm._ 498 (2), 352-358 (2018).
